### DISCUSSION TITLE
What is useEffect

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#22] [Chapter3]
Most of people I met who works by using React, They were thinking **`useEffect` as `class component` life cycles. But actually is not.** It looks like can implement `componentDidMount`, `componentUpdated` and `componentWillUnMount`. Yeah It's really looks so. But looks similar doesn't mean the same. If you one who met I or who thought like I met, let's go deep inside.

> Why most of people misunderstand of using `useEffect`? Why they thought it's just like `class component` life cycles? `useEffect` is ran in 2 conditions.
1. after rendering is done
2. before component is dismiss
> I didn't say `state` updated condition. You might wonder why. When `state` is changed. Condition 1 and 2 are accepted. I think it's ok to treat re-render process is unmount then mount with new `state`. But you should know condition 2 is excuted just before condition 1 with old `state`. It means it ran after rendering. It's just React's principal. So the process goes on like
```
first rendering - run the codes from up to down. except useEffect.
rendering done - run useEffect
re-rendering - run the codes from up to down.
re-rendering done - run useEffect but clean up function is exist, then run it first with old state.
```
> So I didn't tell you update condition. When `state` updated, both conditions are accepted. But what is clean up function? And is all of `state` change accept condition? First, clean up function is just a function which returned from inside of `useEffect`. It looks like
```
useEffect(() => {
.
.
.
  return () => {
    ...
  };
},[someState]);
```
> Inside of `useEffect` a function is returned that's `clean up function`. And it's called like I said. And do you see the 'someState'? That is the `state` which accept condition. If you want some `state` accept the condition so bring the `useEffect`, then put it inside of '[]'. It will work.

#### Finally, I'd like to talk about when is good to use `useEffect`. In my opinion, asynchronous api call is good to use inside of `useEffect`. After rendering is done is good timing to api call. Or add event listener for something. In this case, you have to remove it too right? Add remove listener function at the clean up function is very good usage.

**Comment by chaminju927** [comment]
![image](https://github.com/honaseung/react-deep-dive/assets/118426671/25a3e339-950f-4c8e-9f33-933da7bb004f)
can you explain more in detail - how change of state affects these two conditions?

**Reply by honaseung** [reply]
When re-rendering is happened, component is dismissed first. Then make it once again. So clean up function is ran first. Then the other codes in `useEffect` is ran. But don't forget! It's helping idea for understading. I'm not sure it's just same as real.
